### PR TITLE
Fix NBSP support. Closes: #834.

### LINKED ISSUE #834: Non-breaking space support in `multi_cell()` and `write_html()`
**Error details**

Incorrect support for non-breaking space (NBSP, hard-space, 0x00a0).

Non-breaking space prevents line breaking and it works fine.

Non-breaking space is not fixed-width at all. I know some old systems still use fixed-width NBSP but all modern browsers not.

NBSP (as Unicode character or `&nbsp;` entity) should be variable width space (`SpaceHint`?) in the `FPDF.write_html()`.

The same in the `pdf.multi_cell()`. It could be a compatibility problem if users use it as fixed-width space. Maybe extra option (`fixed_width_nbsp=True`)?

20230703-152836-821x60

**Minimal code**
```python
from fpdf import FPDF

pdf = FPDF(orientation='P', format='A4')
pdf.add_page()
pdf.set_font('helvetica', size=10)
pdf.multi_cell(0, 5, '1\u00a02 aaaaaaaaaaaaaaaaaaaaaqweqwedqweqwe aaaaaaaaaaaaaaaaaaaaaqweqwedqweqweqweqweqweqweqweqwqweqw.</p>')
```

```python
from fpdf import FPDF

pdf = FPDF(orientation='P', format='A4')
pdf.add_page()
pdf.set_font('helvetica', size=10)
pdf.write_html('<p align="justify">1&nbsp;2 aaaaaaaaaaaaaaaaaaaaaqweqwedqweqwe aaaaaaaaaaaaaaaaaaaaaqweqwedqweqweqweqweqweqweqweqwqweqw.</p>')
```

**Environment**
Please provide the following information:
* **Operating System**: Linux Debian Bookworm
* **Python version**: 3.9.17, 3.11.4
* **`fpdf2` version used**: 2.7.4 (pip)

Also check on GitHub `master` branch.

### PR DESCRIPTION
**Pull request by rysson**
Fix NBSP, it's not-break but varible-width space. Fixes #834

**Checklist**:

- [ ] The GitHub pipeline is OK (green), 
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [ ] A unit test is covering the code added / modified by this PR

- [ ] This PR is ready to be merged 

- [ ] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder

- [ ] A mention of the change is present in `CHANGELOG.md`

By submitting this pull request, I confirm that my contribution is made under the terms of the [GNU LGPL 3.0 license](https://github.com/PyFPDF/fpdf2/blob/master/LICENSE).

### CODE REVIEW

**Comment by rysson** [comment]
I need to add tests and changelog.
I'm not sure if docs need be updated.

**Comment by Lucas-C** [comment]
> 

> I need to add tests and changelog. I'm not sure if docs need be updated.

You can just add a mention of your change in `CHANGELOG.md`, that would be enough 😊

**Comment by rysson** [comment]
I don't speak English well, is it enough?
> Fixed:
> - variable-width non-breaking space (NBSP) support [issue #834](https://github.com/PyFPDF/fpdf2/issues/834)

**Comment by gmischler** [comment]
> I'm not sure if docs need be updated.

It might be helpful to mention this eg. in the paragraph about [Character or Word Based Line Wrapping](https://github.com/PyFPDF/fpdf2/blob/40e1b7aa22a4a32fef002f9d5462bcaaa782f3bf/docs/Text.md?plain=1#L34). After the first sentence, you could add something like this:

>>Non-breaking spaces (\U00a0) do not trigger a word wrap, but are otherwise treated exactly as a normal space character.

With a test to actually verify that, this is a very nice little fix!
👍

**Comment by Lucas-C** [comment]
Based on the GitHub Actions execution logs, it seems that your PR is breaking a few existing unit tests:
```
FAILED test/fonts/test_charmap.py...
FAILED test/html/test_html.py::test_html_whitespace_handling
```

File: docs/Text.md [hunk 1]
```diff
@@ -32,8 +32,9 @@ fixed_text = get_display(reshape(some_text))
 ```
 
 ### Character or Word Based Line Wrapping
-By default, `multi_line()` and `write()` will wrap lines based on words, using space characters and soft hyphens as seperators.
-For languages like Chinese and Japanese, that don't usually seperate their words, character based wrapping is more appropriate.
+By default, `multi_line()` and `write()` will wrap lines based on words, using space characters and soft hyphens as separators.
```
Comment (gmischler) [inline docs/Text.md:35]: On that occasion, you could also fix an old mistake, and change the incorrect `multi_line()` in that sentence to `multi_cell()`. :wink:

**Comment by Lucas-C** [comment]
Hi @rysson!

Do you still want to finish this PR? 😊

Otherwise, maybe another contributor could take over

**Comment by rysson** [comment]
@Lucas-C it depends... I'm on holidays now. Then another contributor could finish it faster.
I'll back about Aug, 16.

**Comment by Lucas-C** [comment]
> @Lucas-C it depends... I'm on holidays now. Then another contributor could finish it faster. I'll back about Aug, 16.

No worries, this can definitively wait until the end of August if you wish to finish this! 😊

I'm just going to perform a new release in the meantime.

**Comment by rysson** [comment]
I'm back and I trying to finish this PR.
Everything going wrong, eg. 
- `oscrypto` failed (https://github.com/wbond/oscrypto/issues/78)
- `pytest -vv test/fonts/test_charmap.py test/html/test_html.py` needs `fontTools`, mayme it should be in `test/requirements.txt`?

I'm kidding, nothing to spot me :-P

OK. I've made some fixes. 

@Lucas-C – about unitttest, `test_html_whitespace_handling` is OK IMHO, `test_charmap.py` looks the same, but files differs a bit. Could you look?

**Comment by Lucas-C** [comment]
Hi @rysson

Thank you for getting back to this 😊👍

I haven't reviewed all your changes, but only based on the ones in `CHANGELOG.md`, I think you have _rebase_ issues:
your branch / Pull Request contains many changes in this file that make no sense in this context.

Could you please fix this?
If you cannot manage to properly rebase your branch,
it may be worth creating a new "clean" branch with your patch,
and open a new Pull Request.

**Comment by Lucas-C** [comment]
Hi @rysson

Have you been able to get back to this?

Do you need with the `rebase` or with anything else?

**Comment by rysson** [comment]
> Hi @rysson
> Have you been able to get back to this?

Oh, I've forgot, I'm very sorry.  
Live it too short a little, too many tasks ;-)

Ok, I've used cherry-pick for patch and I've generate same test PDF.
BUT, I have a few more test failed:
```
test/html/test_html.py::test_html_features FAILED
test/html/test_html.py::test_html_customize_ul FAILED
test/html/test_html.py::test_html_description FAILED
test/html/test_html.py::test_html_HTMLMixin_deprecation_warning FAILED
```

It's seems, that space (`20`) is replaced with nbsp (`a0`), for example:
```
E             +  b'BT 31.18 779.94 Td (          description details ) Tj ET',
E             -  b'BT 31.18 779.94 Td <a0a0a0a0a0a0a0a0a0a06465736372697074696f6e2064657461696c'
```

Could you look into those test and tell me if I should generate those PDF too?

**Comment by rysson** [comment]
OK, I generated PDF, it look good IMHO.

I have another test failed:
```python
        with time_execution() as duration:
            build_pdf_with_big_images()
>       assert duration.seconds > 0.3
E       assert 0.2659818779939087 > 0.3
E        +  where 0.2659818779939087 = namespace(seconds=0.2659818779939087).seconds
```

Maybe build server be a little bit slower and test passes.

**Comment by Lucas-C** [comment]
> OK, I generated PDF, they look good IMHO.
> 
> I have another test failed:
> 
> ```python
>         with time_execution() as duration:
>             build_pdf_with_big_images()
> >       assert duration.seconds > 0.3
> E       assert 0.2659818779939087 > 0.3
> E        +  where 0.2659818779939087 = namespace(seconds=0.2659818779939087).seconds
> ```
> 
> Maybe build server be a little bit slower and test passes.

Yes, this is a recurring issue tracked in #923
I'm going to bump this threshold to 22s, but simply retrying your pipeline execution should be enough to have it pass for now.

**Comment by Lucas-C** [comment]
This PR looks good to me 🙂

@gmischler would you like to get a look at it before we merge?

**Comment by gmischler** [comment]
> @gmischler would you like to get a look at it before we merge?

Looks good to me. Let's send it.

**Pull request merged**